webapp-django/crashstats/crashstats/models.py

```diff
@@ -32,15 +31,64 @@
 
 from django.conf import settings
 from django.core.cache import cache
-from django.utils.encoding import iri_to_uri
+from django.db import models
 from django.template.defaultfilters import slugify
+from django.utils.encoding import iri_to_uri
 
 from crashstats.base.utils import requests_retry_session
 
 
 logger = logging.getLogger('crashstats.models')
 
 
+# Django models first
+
+
+class GraphicsDeviceManager(models.Manager):
+    def get_pairs(self, adapter_hexes, vendor_hexes):
+        """Return dict where each tuple of (adapter_hex, vendor_hex)
+        corresponds to a (adapter_name, vendor_name) pair
+
+        """
+        assert len(adapter_hexes) == len(vendor_hexes)
+        names = {}
+        for i, adapter_hex in enumerate(adapter_hexes):
+            cache_key = (
+                'graphics_adapters:' + adapter_hex + ':' + vendor_hexes[i]
+            )
+            key = (adapter_hex, vendor_hexes[i])
+
+            name_pair = cache.get(cache_key)
+
+            if name_pair is None:
+                try:
+                    obj = self.get(adapter_hex=adapter_hex, vendor_hex=vendor_hexes[i])
+                except self.model.DoesNotExist:
+                    continue
+
+                name_pair = (obj.adapter_name, obj.vendor_name)
+                cache.set(cache_key, name_pair, 60 * 60 * 24)
```

Comment:
This reimplementation is simpler than the original. It doesn't do any batching of queries. It (probably) isn't caching things helpfully. But it's a lot easier to read and I figured I'd see how it fared and optimize it later if need be.

As a side note, I'm pretty sure it's hierarchical vendor -> adapter. If so, then we might want to additionally do a query for just the vendor and show at least the vendor name even if there's no vendor/adapter in the db.

Things to think about another day.